### PR TITLE
Scalacheck for aggregates and projections

### PR DESCRIPTION
**Pull request by svalaskevicius**
to cover #13

### CODE REVIEW

File: src/test/scala/CounterSpecification.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      }
+      def nextState(state: State) = state.copy(counter = state.counter + 1)
+      def preCondition(state: State) = state.counter >= 0
+      def postCondition(state: State, success: Boolean) = success
+    }
+
+    case object Decrement extends UnitCommand {
+      def run(sut: Sut) = sut.synchronized {
+        val action = sut.runner.db(agg, Counter.counterAggregate.handleCommand(Counter.Decrement))
+        sut.runner = sut.runner.run(action) match {
```
Comment (svalaskevicius) [inline src/test/scala/CounterSpecification.scala]: not particularly nice... I wonder if there is a better way?

File: src/test/scala/CounterSpecification.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import Cqrs.Aggregate._
+import Cqrs.Database.EventSerialisation
+import Domain.Counter
+import org.scalacheck.{Properties, Gen}
+
+object CounterSpecification extends Properties("Counter") {
+
+  property("fails when below zero") = CounterSpec.property()
+
+  private object CounterSpec extends AggregateCommands[Counter.Event, Counter.flow.StateData, Int] {
```
Comment (svalaskevicius) [inline src/test/scala/CounterSpecification.scala]: @Fristi hi! 

what do you think about this kind of test for an aggregate? I'll create a similar one for the door agg.

projections, however, will need something more clever than this...

is there an alternative approach you can think of?

File: src/test/scala/DoorSpecification.scala [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import java.awt.print.Book
```
Comment (Fristi) [inline src/test/scala/DoorSpecification.scala]: Book? :smile:

Comment (svalaskevicius) [inline src/test/scala/DoorSpecification.scala]: Oops :) copy paste err.

File: src/test/scala/CounterSpecification.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      Gen.oneOf(
+        Gen.const(Increment),
+        Gen.const(Decrement)
+      )
+
+    override def evtSerializer = implicitly[EventSerialisation[Counter.Event]]
+
+    case object Increment extends AggregateCommand {
+      def commandActions = Counter.CounterAggregate.handleCommand(Counter.Increment)
+      def nextState(state: State) = state + 1
```
Comment (Fristi) [inline src/test/scala/CounterSpecification.scala]: Hmm, isn't this duplicate logic? If a command passes, you should raise an event which will update the state which is also encoded in the domain logic. Or am I mistaken?

File: src/test/scala/CounterSpecification.scala (hunk 2 above)
Comment (Fristi) [inline src/test/scala/CounterSpecification.scala]: Sorry for the late response! Been busy lately with moving/new job :smile:

Comment (svalaskevicius) [inline src/test/scala/CounterSpecification.scala]: No problem! 

Congrats on a new job :)

File: src/test/scala/CounterSpecification.scala (hunk 4 above)
Comment (svalaskevicius) [inline src/test/scala/CounterSpecification.scala]: Yes... After implementing this commands yest model I think it's not as useful as I initially  thought.. 

Simple quickcheck cases for specific domain should be much clearer.. What do you think?

File: src/test/scala/DoorSpecification.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    case class Lock(key: String) extends AggregateCommand {
+      def commandActions = Door.DoorAggregate.handleCommand(Door.Lock(key))
+      def nextState(state: State) = if (state == TestState.Closed) TestState.Locked(key) else state
+      def postCondition(state: State, success: Boolean) = success == (state == TestState.Closed)
+    }
+
+    case class Unlock(key: String) extends AggregateCommand {
+      def commandActions = Door.DoorAggregate.handleCommand(Door.Unlock(key))
+      def nextState(state: State) = if (state == TestState.Locked(key)) TestState.Closed else state
+      def postCondition(state: State, success: Boolean) = success == (state == TestState.Locked(key))
```
Comment (Fristi) [inline src/test/scala/DoorSpecification.scala]: A aggregate is a async boundary which must guard these invariants which is captured nicely here. This case is a really interesting invariant property check as it needs to unlock the door when you give the right key. Great job captureing that here!

Though I have some observations :smile: 

`postCondition` accepts a boolean `success` parameter, but where does that come from? This boolean is captured outside this function. Wouldn't it be simpler to omit this boolean and evaluate if `success` is equal to the expression you give in this case: `state == TestState.Locked(key)`

**Comment by Fristi** [comment]
I really like this direction! Especially the invariant property tests, great job! I have made some comments, see the diff

**Pull request merged**